Make the SBT plugin more consistent with the Maven plugin and follow SBT plugin best practices

stevedlawrence opened the pull request:
- Modify the settings to be prefixed with the 'rat' namespace so as to
  prevent potential key collisions. Also rename some settings so that
  they are the same as the rat maven plugin for consistency. Rename
  auditReport and auditCheck to ratReport and ratCheck, since there may
  be other 'audit' tasks out there that might conflict.
- Rather than running rat checks on just unmanaged sources--which leaves
  out lots of files including resources, sbt configuration files, and
  files that sbt doesn't know about--recurse down the entire source
  directory and perform a check on all files. This also means that there
  is really no need for a separate configuration to run on compile vs
  test, so the Audit config is removed and Compile/Test config stuff is
  removed.
- The above change will now check lots of files that it shouldn't, like SCM
  files and sbt generated files. So add a new setting
  (ratParseSCMIgnoresAsExcludes) to parse SCM ignore files so that they
  can be excluded. Additionally, add a new setting (ratExclude) to
  provide specific files that should be ignored from a rat check. This
  is useful to exclude files that are tracked via SCM, but cannot include
  the Apache license header, for example test files.
- Add setting (ratReportStyle) to configure the report style. Users
  probably don't need to output two different types of reports. Rather
  than always outputting as both text and ascii doc, provide a setting
  to specify which style is preferred and only output that style. Modify
  tests so one uses txt and the other uses adoc.
- Remove the use of maxErrors. That key is really for just how many
  compile errors to display, not really how many errors constitutes a
  fatal error, which is sortof how this plugin was using it. Consistent
  with the maven plugin, it should be an error if _any_ unapproved
  licenses found, so maxErrors really isn't necessary.

Note that the ScmIngoreParser.java file is copied from the rat maven
plugin with minor modifications to remove logging support. Ideally that
function would be moved to rat-core since really all plugins could
benefit from the ability to use SCM ignores.

stevedlawrence committed:
Make the SBT plugin more consistent with the Maven plugin and follow SBT plugin best practices

- Modify the settings to be prefixed with the 'rat' namespace so as to
  prevent potential key collisions. Also rename some settings so that
  they are the same as the rat maven plugin for consistency. Rename
  auditReport and auditCheck to ratReport and ratCheck, since there may
  be other 'audit' tasks out there that might conflict.
- Rather than running rat checks on just unmanaged sources--which leaves
  out lots of files including resources, sbt configuration files, and
  files that sbt doesn't know about--recurse down the entire source
  directory and perform a check on all files. This also means that there
  is really no need for a separate configuration to run on compile vs
  test, so the Audit config is removed and Compile/Test config stuff is
  removed.
- The above change will now check lots of files that it shouldn't, like SCM
  files and sbt generated files. So add a new setting
  (ratParseSCMIgnoresAsExcludes) to parse SCM ignore files so that they
  can be excluded. Additionally, add a new setting (ratExclude) to
  provide specific files that should be ignored from a rat check. This
  is useful to exclude files that are tracked via SCM, but cannot include
  the Apache license header, for example test files.
- Add setting (ratReportStyle) to configure the report style. Users
  probably don't need to output two different types of reports. Rather
  than always outputting as both text and ascii doc, provide a setting
  to specify which style is preferred and only output that style. Modify
  tests so one uses txt and the other uses adoc.
- Remove the use of maxErrors. That key is really for just how many
  compile errors to display, not really how many errors constitutes a
  fatal error, which is sortof how this plugin was using it. Consistent
  with the maven plugin, it should be an error if _any_ unapproved
  licenses found, so maxErrors really isn't necessary.

Note that the ScmIngoreParser.java file is copied from the rat maven
plugin with minor modifications to remove logging support. Ideally that
function would be moved to rat-core since really all plugins could
benefit from the ability to use SCM ignores.

stevedlawrence committed:
ratExlucdes that are just a single file will match any file/directory with that name

Also, disable aggregation of ratCheck and ratReport, those should only
be run on the root project and it will recurse down the entire directory
tree

stevedlawrence committed:
Add ratTarget option to change the output rat report file

Defaults to "rat." + ratReportStyle

Also modifies the ratCheck exception to display this path rather than
always displaying "rat.txt"

stevedlawrence commented:
Added a few more changes to this pull request. The first is to work better with projects that use subprojects by disabling aggregation. The idea is that this should only be run from the root project. I think it might be an interesting enhancement in the future to better handle submodules, but makes things a bit more complex, and I'm not sure there's a whole lot of benefit. The other changes are to better match SCM ignores, and to add a ratTarget setting to configure where the output report goes to.

These changes have been tested and work well with Apache Daffodil.

jvz commented:
Oh I'm so sorry I missed this! I've been having issues with GitHub notifications lately.

jvz commented on src/main/java/org/apache/rat/mp/util/ScmIgnoreParser.java:
```diff
@@ -0,0 +1,113 @@
+ * Helper to parse SCM ignore files to add entries as excludes during RAT runs.
+ * Since we log errors it needs to reside inside of the maven plugin.
+ */
+public final class ScmIgnoreParser {
```
Is this class available from rat-core? If not, it'd be nice to rewrite this in Scala, though I can handle that later.

stevedlawrence replied:
Unfortunately, it packaged in apache-rat-plugin.jar. I thought about depending on that jar, but it pulls in tons of unnecessary dependencies. It would be better to either rewrite it in scala, or just try to convince the Apache Rat devs to move it into apache-core. That latter seems like a better idea since it really would be useful to all users of apache-core, but maybe there are reasons for having it in the plugin.

jvz commented on src/main/scala/org/musigma/sbt/rat/SbtRatPlugin.scala:
```diff
@@ -37,93 +39,163 @@ object SbtRatPlugin extends AutoPlugin {
   object autoImport {
-    // FIXME: this should probably be a task, not a configuration (e.g., packageBin)
-    val Audit = config("audit")
+    val ratCheck = taskKey[Unit]("Performs a release audit check")
```
Thanks for handling this. I was originally following a pattern from sbt-site, but it didn't seem to be useful.

jvz commented on src/main/scala/org/musigma/sbt/rat/SbtRatPlugin.scala:
```diff
@@ -37,93 +39,163 @@ object SbtRatPlugin extends AutoPlugin {
 
-    val auditCheck = taskKey[Unit]("Performs a release audit check")
-    val auditReport = taskKey[AuditReport]("Generates a release audit report")
+    type AuditReport = ClaimStatistic
```
If you're renaming the tasks, you should rename this type alias to `RatReport` as well.

stevedlawrence replied:
Push a fix for this change

jvz commented on src/main/scala/org/musigma/sbt/rat/SbtRatPlugin.scala:
```diff
@@ -37,93 +39,163 @@ object SbtRatPlugin extends AutoPlugin {
+    val ratLicenses = settingKey[Seq[(String, String, String)]]("Extra license to match--each Seq item is a tuple of (category, name, license text)")
 
-    type AuditReport = ClaimStatistic
+    val ratExcludes = settingKey[Seq[File]]("Files to exlucde from audit checks, relative to the baseDirectory")
```
This type of thing would normally be done by `excludes in ratReport` or `includes in ratReport`, though I'm not sure how to implement that. I'd prefer consistency with most other sbt plugins here.

stevedlawrence replied:
I looked into this and couldn't figure out how to do it in a good way.  In order for "excludes in ratReport" to work, it looks like the ``val excludes = `` setting would need to be defined in the autoImport object. But that means it is auto imported into the global namespace. And a setting like ``excludes`` seems pretty generic and should probably be namespaced somehow. So if we kept the "rat" prefix, the build.sbt config would like like ``ratExcludes in ratReport := ...`` which feels a little odd to me.

Maybe something like ``excludes in ratReport`` is standard if ``excludes`` is a an already defined and standard Sbt key, and ratReport would want to override it? The [Sbt Plugin Best Practices Page](https://www.scala-sbt.org/1.0/docs/Plugins-Best-Practices.html) does something like that with ``sources in obfuscate``, but sources is already defined, so it's slightly different than 'excludes'. Other settings that are specific to their Obfucsate example are not scoped and and just prefixed with ``obfuscate``. 

But that's just my guess by reading the documentation. I'm not familiar with other plugins that do something like this (or many plugins in general). If you know of any, point them out and I can take a look at how they do it.

jvz replied:
Looks like I mixed up sbt and Maven. I meant `excludeFilter` and `includeFilter` which are defined in `sbt.Keys`.

stevedlawrence replied:
Ah, ok. My only thing against excludeFilter is that it's a pain to use, especially if you want to exclude specific paths instead of just names (which is what we want in Daffodil). However, this patchset actually ends up creating a FileFilter based on the ratExcludes value, so maybe that can be a good default for excludeFilter in ratReport, and if someone wants something different then they can override it? I think that would be fairly straightforward to implement.

jvz replied:
A good default for excludeFilter makes sense to me. The user can override the filter by either replacing it or appending to it, so it seems consistent with sbt practices.

stevedlawrence replied:
Ok, pushed three new commits. The first is unrelated but something I had hanging around. It gives a cleaner output when ratCheck finds unapproved licenses (i.e. no meaningless stack trace). The second just moves the existing functionality into objects to follow the best practices--makes things cleaner and easier to support the last commit. The last commit adds support for includeFilter and excludeFilter, with reasonable defaults that can be overridden or appended to if needed.

jvz reviewed (changes requested):
Overall looking pretty good. Just a couple changes and we should be good to go!

stevedlawrence committed:
Rename AuditReport to RatReport for consistency

Also import the Apache RatReport as ApacheRatReport to remove the name
conflict

stevedlawrence committed:
Change how ratCheck reports failures

Change the UnapprovedLicense exception to be a
FeedbackProvidedException, preventing the output of a stacktrace that is
meaningless to a user. Instead, use log.error to output the error
message and throw this new exception.

stevedlawrence committed:
Update to meet SBT plugin best practices

Move the ratReport and ratCheck functionality into objects with an apply
method. This makes things a little clearner and follows the best
practices. No actual functional changes.

stevedlawrence committed:
Use "includeFilter in ratReport" and "excludeFilter in ratReport" to determine which files check

- "includeFilter in ratReport" defaults to including all files via the
  AllPassFilter.
- "excludeFilter in ratReport" defaults to excluding all files defined
  in ratExcludes and SCM ignore files if ratParseSCMIgnoresAsExcludes is
  true.
- Both settings can be easily overridden or appended to if a user needs
  extra functionality.
- Modify test to show usage of appending to excludeFilter in ratReport

jvz commented on src/sbt-test/sbt-rat/apache-headers/build.sbt:
```diff
@@ -1,2 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
I mean technically we're developing this on my GitHub instead of at Apache. Though I believe you'll need to update the NOTICE/COPYRIGHT files if you're bringing in outside code (like the .java file earlier).

Edit: the NOTICE file already had the note about Apache software inclusion, so that's alright.

stevedlawrence replied:
Yep, I originally intended this to be an RFC and figured this this file would be moved to rat-core and wouldn't be included anymore, but that might take some effort/time. I'll update the licenses appropriately.

jvz replied:
I'm hoping to work on integrating this upstream to Creadur at some point, so we'll be granting it back to them anyways.

stevedlawrence replied:
Actually, the NOTICE of this plugin currently has the standard ASF notice:
```
This product includes software developed at
The Apache Software Foundation (https://www.apache.org/)
```
My understanding is that since the both pieces of software are under the Apache license and since both have the standard "developed at ASF" notice without anything extra, then it isn't necessary to list all the the notice again. So maybe we don't need to modify anything for this file?

jvz replied:
Right, the notice is fine since it has the notice already.

jvz commented on src/sbt-test/sbt-rat/apache-headers/src/main/scala/BSD3.scala:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2011-2013, Someone
```
I'd prefer something more legal. How about `Copyright 1850 Abraham Lincoln` or something expired? ;)

stevedlawrence replied:
IANAL, but claiming something is copyright someone else doesn't feel right. Maybe we could switch to a file with a different license, maybe Creative Commons would have less issues? The test doesn't really care what the license is.

jvz replied:
Ah, well, perhaps a real copyright then? Or at least match the actual year. The file is trivial enough that I doubt copyright would actually apply to it besides the license itself.

stevedlawrence replied:
Will update to a real copyright. Agreed that copyright probably doesn't even really apply since this is so trivial.

stevedlawrence committed:
Use an actual copyright for the BSD3 file

Note that this is a trivial test file just to make sure that the plugin
can match the BSD3 license.

jvz commented:
Make sure to bump the minor version in both version.sbt and the readme. Once I merge this, it'll release automatically.

stevedlawrence committed:
Bump version to 0.5.2

jvz commented:
I meant 0.6.0! ;)

The API is changing a bit here, so we'll go with the minor instead of patch version. Can likely make a 1.0 release soon now that we're converging on the base feature set.

stevedlawrence committed:
Bump to 0.6.0 due to API changes

jvz commented:
Thanks for all the help! Now let's hope this plugin release works like it used to since I haven't touched the config in a couple months. ;)

stevedlawrence commented:
Thanks for reviewing it!

jvz commented:
Once version 0.6.0 is published to Maven Central, let me know if you have any issues.
